### PR TITLE
✨ Add support for minify-replace babel plugin

### PR DESCRIPTION
**Pull request by jonathantyng-amp**
Adds support for the [`minify-replace`](https://www.npmjs.com/package/babel-plugin-minify-replace) plugin in the babel build process.

This is needed to change experimental flags in the code during build time. Experiment flags will be tracked in `build-system/global-configs/experiments-config.json` and will default to `false`. Passing in `--define=SOME_EXPERIMENT_CONSTANT` will set it to true or `--define=SOME_EXPERIMENT_CONSTANT=true` will set it to some value.

Dependent on https://github.com/ampproject/amphtml/pull/22839

### CODE REVIEW

File: build-system/build.conf.js [hunk 1]
```diff
@@ -26,9 +29,59 @@ const defaultPlugins = [
...
+    {replacements},
+  ];
+
+  return replacePlugin;
+}
+
 module.exports = {
   plugins({isEsmBuild, isForTesting, isSinglePass}) {
-    let pluginsToApply = defaultPlugins;
+    let pluginsToApply = defaultPlugins.slice();
```
Comment (jonathantyng-amp) [inline build-system/build.conf.js]: I was getting an error locally where `babel-plugin-transform-amp-asserts` was duplicated in this array causing babel to error out. `pluginsToApply` is initially a reference to `defaultPlugins` and pushing this plugin caused subsequent uses of `defaultPlugins` to have this duplicate. Not sure why it was passing the PR build and failing locally but creating a copy seems to fix this.

**Comment by jonathantyng-amp** [comment]
cc @erwinmombay @zhouyx for review :)

**Comment by erwinmombay** [comment]
I think we can proceed with this for now. (just to unblock work)

@kristoferbaxter would like for us to create our own plugin that does this work. It'll give the ads team an opportunity to write a babel plugin (which is good since the team in general will probably be writing more babel plugins).

File: build-system/build.conf.js [hunk 2]
```diff
@@ -26,9 +29,59 @@ const defaultPlugins = [
...
+  // default each experiment flag constant to false
+  Object.keys(experimentsConfig).forEach(experiment => {
+    const experimentDefine = experimentsConfig[experiment]['define'];
+    if (experimentDefine) {
+      replacements.push(createReplacement(experimentDefine + '=false'));
+    }
+  });
+
+  // override define values from passed in flags
+  if (Array.isArray(argv.define)) {
```
Comment (zhouyx) [inline build-system/build.conf.js]: Does this mean that we support compiling multiple experiments at the same time? I don't find that feature to be very useful. A bit confusing instead. Can we only support one value at each build? Thanks.

File: build-system/build.conf.js [hunk 3]
```diff
@@ -26,9 +29,59 @@ const defaultPlugins = [
...
+    argv.define.forEach(defineStr => {
+      replacements.push(createReplacement(defineStr));
+    });
+  } else if (argv.define) {
+    replacements.push(createReplacement(argv.define));
```
Comment (zhouyx) [inline build-system/build.conf.js]: To confirm two replacements item with the same identifierName, the ladder one override the previous one right?

File: build-system/global-configs/experiments-config.json [hunk 4]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (zhouyx) [inline build-system/global-configs/experiments-config.json]: Could you please add comments on what the experiment config template should look like? Thanks.

File: package.json [hunk 5]
```diff
@@ -49,6 +49,7 @@
     "babel-eslint": "10.0.2",
     "babel-plugin-filter-imports": "3.0.0",
     "babel-plugin-istanbul": "5.1.4",
+    "babel-plugin-minify-replace": "^0.5.0",
```
Comment (rsimha) [inline package.json]: Prefer exact versions. You can do this by using `--exact` with your `yarn add` command.

**Comment by zhouyx** [comment]
Thank you @erwinmombay for the info. We can definitely aim to create our own babel plugin and not rely on closure to DCE code in future. 

Given we want to make the plugin generic for all future experiments. We want to remove code not case by case. @jonathantyng-amp proposed maybe we can use code comment and move code with some annotation? Another way I am thinking is to wrap to be removed code in function with a special name prefix. 

@erwinmombay and @kristoferbaxter What do you think the custom babel plugin code removal logic should be like? Any ideas on a generic way to remove code?  Thank you

File: build-system/build.conf.js [hunk 6]
```diff
@@ -26,9 +29,59 @@ const defaultPlugins = [
...
 ];
 
+/**
+ * @return {Array<string|Object>} the minify-replace plugin options that can be
+ * pushed into the babel plugins array
+ */
+function getReplacePlugin() {
+  /**
+   * @param {string} defineStr the define flag to parse
+   * @return {Object} replacement options used by minify-replace plugin
```
Comment (erwinmombay) [inline build-system/build.conf.js]: `!Object`

File: build-system/build.conf.js (hunk 2 above)
Comment (jonathantyng-amp) [reply build-system/build.conf.js]: Done, will now throw an error if more than one `define` flag is passed in

File: build-system/build.conf.js (hunk 3 above)
Comment (jonathantyng-amp) [reply build-system/build.conf.js]: Yes! Added a check to not have duplicates just to make this clearer

File: build-system/global-configs/experiments-config.json (hunk 4 above)
Comment (jonathantyng-amp) [reply build-system/global-configs/experiments-config.json]: Done!

**Comment by jonathantyng-amp** [comment]
Ready for another review!

**Comment by erwinmombay** [comment]
@zhouyx lets talk offline to flesh it out! thanks

File: build-system/build.conf.js [hunk 7]
```diff
@@ -26,9 +29,69 @@ const defaultPlugins = [
...
+      identifierName,
+      replacement: {
+        type: 'booleanLiteral',
+        value,
+      },
+    };
+  }
+
+  const replacements = [];
+  const defineFlag = argv.define;
```
Comment (erwinmombay) [inline build-system/build.conf.js]: did we settle on `define` as the flag name?

File: build-system/build.conf.js (hunk 1 above)
Comment (erwinmombay) [reply build-system/build.conf.js]: add `slice(0)` just to be explicit or do `[...defaultPlugins]`

File: build-system/build.conf.js (hunk 7 above)
Comment (jonathantyng-amp) [reply build-system/build.conf.js]: nope the name isn't settled, any suggestions?

Comment (jonathantyng-amp) [reply build-system/build.conf.js]: discussed offline, settled on `defineExperimentConstant`

File: build-system/build.conf.js [hunk 8]
```diff
@@ -26,9 +29,70 @@ const defaultPlugins = [
...
+ * @return {Array<string|Object>} the minify-replace plugin options that can be
+ * pushed into the babel plugins array
+ */
+function getReplacePlugin() {
+  /**
+   * @param {string} defineStr the define flag to parse
+   * @return {!Object} replacement options used by minify-replace plugin
+   */
+  function createReplacement(defineStr) {
+    const strSplit = defineStr.split('=');
```
Comment (zhouyx) [inline build-system/build.conf.js]: Is this to support `--defineExperimentConstant=EXP_A=true,EXP_B=false`. Just curious does `=` works well in the argv?

Comment (jonathantyng-amp) [reply build-system/build.conf.js]: Yes, this is to support `--defineExperimentConstant=EXP_A=true`. So you can set EXP_A to any value

**Pull request merged**